### DISCUSSION TITLE
Implicit model binding doesn't work with full-page component routes and authorization via ->middleware('can:')

### DISCUSSION THREAD

**Discussion by ryanmortier** [livewire/livewire#3106] [7. Bugs (archived)] [answered]
### Description

When using full-page components and authorization via route middleware, implicit model binding does not work and the route parameter is passed to the gate as a string.

### Exact steps to reproduce

Define a gate:

```php
Gate::define('view-order', function (User $user, Order $order) {
    dd($order); // this is a string when it should be an order object
});
```

Define a full-page component route with middleware:

```php
Route::get('/orders/{order}', \App\Http\Livewire\OrderShow::class)->middleware('can:view-order,order');
```

Visit in your browser (an example):
```
http://127.0.0.1:8000/orders/1
```

The string literal "1" is passed to the gate rather than the model that should be resolved via implicit model binding.

**Context**
- Livewire version: 2.3.18
- Laravel version: 8.28.1
- Browser: Chrome

**Comment by erikgaal** [comment]
This functionality in Laravel relies on the route parameter to be bound by the https://github.com/laravel/framework/blob/8.x/src/Illuminate/Routing/ImplicitRouteBinding.php class. That class relies on the https://github.com/laravel/framework/blob/8.x/src/Illuminate/Routing/RouteSignatureParameters.php to extract the signature parameters from the action.

Unfortunately, it cannot extract the parameters from the Livewire component, because the action is actually the `__invoke` method. Of course, this does not have the route models as its parameters.

If we instead could attach an extra piece of middleware to each Livewire Component that is a route action that runs before the `Authorize` middleware and resolves the correct paramters, this might work.

**Comment by erikgaal** [comment, marked as answer]
Alternatively, you can explicitly bind the model to the route parameter using `Route::model('order', Order::class)`.

**Reply by rcubitto** [reply]
This will break explicit model binding `/orders/{order:uuid}`

**Comment by georgeboot** [comment]
We added the following to our component to bypass this limitation:

```php
<?php

namespace App\Http\Controllers;

use App\Models\TestModel;
use Illuminate\Contracts\Container\Container;
use Illuminate\Routing\Route;
use Livewire\Component;

class TestComponent extends Component
{
    public function __invoke(Container $container, Route $route, ?TestModel $testModel = null)
    {
        return parent::__invoke($container, $route);
    }

     // ...
}
```

This will make sure that PHP's reflection sees the expected model on the routes action (`App\Http\Controllers\TestComponent@__call`).

**Reply by rcubitto** [reply]
This worked for me, thank you!

**Comment by redgluten** [comment]
I’m not exactly sure how @calebporzio feels about the proposed solutions but this should be at least mentioned in the docs in the [“Full-Page Components” section](https://laravel-livewire.com/docs/2.x/rendering-components#page-components). I would gladly start a P.R with a link to this discussion unless another solution is preferred?
